src/python/pants/rules/core/fastlist.py

```diff
@@ -4,14 +4,91 @@
 
 from __future__ import absolute_import, division, print_function, unicode_literals
 
+from pants.base.specs import Specs
 from pants.engine.addressable import BuildFileAddresses
 from pants.engine.console import Console
-from pants.engine.rules import console_rule
-from pants.engine.selectors import Select
+from pants.engine.legacy.graph import HydratedTargets
+from pants.engine.rules import console_rule, optionable_rule
+from pants.engine.selectors import Get, Select
+from pants.subsystem.subsystem import Subsystem
 
 
-@console_rule('list', [Select(Console), Select(BuildFileAddresses)])
-def fast_list(console, addresses):
+class ListOptions(Subsystem):
+  """Lists all targets matching the target specs.
+
+  If no targets are specified, lists all targets in the workspace.
+  """
+
+  # NB: This option scope is temporary: a followup to #6880 will replace the v1 list goal and rename
+  # this scope.
+  options_scope = 'fastlist'
+
+  @classmethod
+  def register_options(cls, register):
+    super(ListOptions, cls).register_options(register)
+    register('--provides', type=bool,
+             help='List only targets that provide an artifact, displaying the columns specified by '
+                  '--provides-columns.')
+    register('--provides-columns', default='address,artifact_id',
+             help='Display these columns when --provides is specified. Available columns are: '
+                  'address, artifact_id, repo_name, repo_url, push_db_basedir')
+    register('--documented', type=bool,
+             help='Print only targets that are documented with a description.')
+
+
+@console_rule('list', [Select(Console), Select(ListOptions), Select(Specs)])
+def fast_list(console, options, specs):
   """A fast variant of `./pants list` with a reduced feature set."""
```

Comment:
Can you remove the "reduced feature set" caveat? I don't see anything missing here now.

Reply:
It's missing some of the options that used to be inherited from `ConsoleTask`: that will be resolved in #6880.